File: Keep_Stock_Web/models/clientes.py
from core.crud_base import CrudBase
from core import database
from core.database import conectar_supabase
from core.validator import Validator
 
class Cliente(CrudBase):
    table = "clientes"
    primary_key = "id_cliente" 
    fields = [
        "nome",
        "cpf_cnpj",
        "telefone",
        "email",
        "codigo_postal",
        "imagem_url"
    ]

    def __init__(self, nome, cpf_cnpj, telefone, email, codigo_postal, imagem_url=None, id=None, id_cliente=None, **kwargs):
        self.nome = nome
        self.cpf_cnpj = cpf_cnpj
        self.telefone = telefone
        self.email = email
        self.codigo_postal = codigo_postal
        self.imagem_url = imagem_url
        self.id = id or id_cliente
        self.id_cliente = self.id
        self.db = conectar_supabase()  # Pega o cliente do Supabase
 
    @classmethod
    def find_all(cls, order_by=None):
        # Sobrescreve find_all para usar id_cliente como chave primária
        if order_by is None:
            order_by = cls.primary_key
        supabase = conectar_supabase()
        response = supabase.table(cls.table).select("*").order(order_by).execute()
        return response.data

    @classmethod
    def find_by_id(cls, id):
        # Sobrescreve find_by_id para usar id_cliente como chave primária
        supabase = conectar_supabase()
        response = supabase.table(cls.table).select("*").eq(cls.primary_key, id).execute()
        return response.data[0] if response.data else None

    @classmethod
    def delete(cls, id):
        # Sobrescreve delete para usar id_cliente como chave primária
        supabase = conectar_supabase()
        response = supabase.table(cls.table).delete().eq(cls.primary_key, id).execute()
        return response.data

    def insert(self):
        # Sobrescreve insert 
        supabase = conectar_supabase()
        dados = {campo: getattr(self, campo) for campo in self.fields}
        response = supabase.table(self.table).insert(dados).execute()
        return response.data

    def update(self, id):
        # Usar update para usar id_cliente como chave primária
        supabase = conectar_supabase()
        dados = {campo: getattr(self, campo) for campo in self.fields}
        response = supabase.table(self.table).update(dados).eq(self.primary_key, id).execute()
        return response.data

    def validate(self):
        erros = [
            Validator.required(self.nome, "nome"),
            Validator.required(self.cpf_cnpj, "cpf/cnpj"),
            Validator.required(self.telefone, "telefone"),
            Validator.required(self.email, "email"),
            Validator.required(self.codigo_postal, "código postal"),
        ]
        return [erro for erro in erros if erro]
                    
    @classmethod
    def has_related_records(cls, id):
        # Verifica se o cliente possui registros relacionados (pedidos, saídas)
        supabase = conectar_supabase()
        try:
            response = supabase.table("pedidos").select("*", count="exact").eq("id_cliente", id).execute()
            if getattr(response, "count", None) and response.count > 0:
                return True

            response = supabase.table("saidas_produtos").select("*", count="exact").eq("id_cliente", id).execute()
            if getattr(response, "count", None) and response.count > 0:
                return True

            return False
        except Exception as e:
            print(f"Erro ao verificar registros relacionados do cliente: {e}")
            return False
                    
    @classmethod
    def safe_delete(cls, id):
        supabase = conectar_supabase()
        # Busca cliente e verifica se ele existe
        cliente = cls.find_by_id(id)
        if not cliente:
            raise ValueError("cliente não encontrado.")
        if cls.has_related_records(id):
            raise ValueError("Não é possível excluir o cliente porque possui tabelas vinculadas.")
        cls.delete(id)

    @classmethod
    def buscar_por_cliente(cls, id_cliente):
        # Busca um cliente no banco e retorna uma instância da classe
        supabase = conectar_supabase()
        response = supabase.table("clientes").select("*").eq(cls.primary_key, id_cliente).execute()

        if response.data:
            d = response.data[0]
            return cls(
                id=d.get('id_cliente'),
                nome=d.get('nome'),
                cpf_cnpj=d.get('cpf_cnpj'),
                telefone=d.get('telefone'),
                email=d.get('email'),
                codigo_postal=d.get('codigo_postal'),
                imagem_url=d.get('imagem_url')
            )
        return None
    
    def inserir_cliente(self):
        # Insere o cliente usando o método da classe base
        supabase = conectar_supabase()
        dados = {campo: getattr(self, campo) for campo in self.fields}
        response = supabase.table(self.table).insert(dados).execute()
        return response.data

File: Mobile_Keep_Stock/src/services/clientes.js
import { criarCrud } from './crudBase';
import { filtrarPorTermo } from './helpers';
import { BUCKET_CLIENTES, deletarImagem, uploadImagem } from './imagemStorage';
import { Validator } from './validator';

// ROTAS DE CLIENTES
// Espelha /clientes e /cliente/salvar, /cliente/atualizar e /cliente/excluir do sistema web

export const clientesCrud = criarCrud('clientes', 'id_cliente');

// Normaliza os campos vindos do formulário (espelha get_cliente_form)
export function montarCliente(dados) {
  return {
    nome: (dados.nome || '').trim(),
    cpf_cnpj: (dados.cpf_cnpj || '').trim(),
    telefone: (dados.telefone || '').trim(),
    email: (dados.email || '').trim(),
    codigo_postal: (dados.codigo_postal || '').trim(),
  };
}

// Espelha Cliente.validate()
export function validarCliente(cliente) {
  const erros = [
    Validator.required(cliente.nome, 'nome'),
    Validator.required(cliente.cpf_cnpj, 'cpf/cnpj'),
    Validator.required(cliente.telefone, 'telefone'),
    Validator.required(cliente.email, 'email'),
    Validator.required(cliente.codigo_postal, 'código postal'),
  ];

  return erros.filter(Boolean);
}

export async function listarClientes(termo = '') {
  const clientes = await clientesCrud.findAll('nome');

  return filtrarPorTermo(clientes, termo, [
    'nome',
    'cpf_cnpj',
    'telefone',
    'email',
    'codigo_postal',
  ]);
}

export async function buscarCliente(id) {
  return clientesCrud.findById(id);
}

export async function salvarCliente(dados, imagem = null) {
  const cliente = montarCliente(dados);
  let imagemEnviada = null;

  try {
    imagemEnviada = imagem ? await uploadImagem(imagem, BUCKET_CLIENTES) : null;
    cliente.imagem_url = imagemEnviada;

    const erros = validarCliente(cliente);

    if (erros.length) throw new Error(erros.join(' '));

    return await clientesCrud.insert(cliente);
  } catch (erro) {
    if (imagemEnviada) await deletarImagem(imagemEnviada, BUCKET_CLIENTES);

    throw erro;
  }
}

export async function atualizarCliente(id, dados, imagem = null) {
  const clienteAtual = await clientesCrud.findById(id);

  if (!clienteAtual) throw new Error('Cliente não encontrado.');

  const cliente = montarCliente(dados);
  cliente.imagem_url = clienteAtual.imagem_url;

  let novaImagem = null;

  try {
    if (imagem) {
      novaImagem = await uploadImagem(imagem, BUCKET_CLIENTES);
      cliente.imagem_url = novaImagem;
    }

    const erros = validarCliente(cliente);

    if (erros.length) throw new Error(erros.join(' '));

    const atualizado = await clientesCrud.update(id, cliente);

    if (novaImagem && clienteAtual.imagem_url) {
      await deletarImagem(clienteAtual.imagem_url, BUCKET_CLIENTES);
    }

    return atualizado;
  } catch (erro) {
    if (novaImagem) await deletarImagem(novaImagem, BUCKET_CLIENTES);

    throw erro;
  }
}

// Verifica se o cliente possui registros relacionados (pedidos, saídas)
export async function clientePossuiVinculos(id) {
  try {
    const pedidos = await criarCrud('pedidos').contarPor('id_cliente', id);

    if (pedidos > 0) return true;

    const saidas = await criarCrud('saidas_produtos').contarPor('id_cliente', id);

    if (saidas > 0) return true;

    return false;
  } catch (erro) {
    console.log(`Erro ao verificar registros relacionados: ${erro}`);
    return false;
  }
}

// Espelha Cliente.safe_delete()
export async function excluirCliente(id) {
  const cliente = await clientesCrud.findById(id);

  if (!cliente) throw new Error('Cliente não encontrado.');

  if (await clientePossuiVinculos(id)) {
    throw new Error(
      'Não é possível excluir o cliente porque possui tabelas vinculadas.'
    );
  }

  await clientesCrud.delete(id);

  if (cliente.imagem_url) await deletarImagem(cliente.imagem_url, BUCKET_CLIENTES);
}
